[20기 문성호] 장바구니 구현

Room9 opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 장바구니 추가 / 삭제 / 수정 / 조회 구현
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- order_status 가 '주문 전'인 status를 가지는 order_id를 참조하도록 장바구니 구현
- 장바구니 추가 시, 이미 추가된 상품인 경우 상품의 갯수 +1 되도록 로직 구현
- 장바구니 내 특정 상품의 갯수 수정 로직 구현
- 장바구니 내 상품 삭제 / 조회 구현
- 하나의 class 내에서 method를 post/ patch / delete / get으로 구분하여 단일 endpoint에 묶이도록 했습니다

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
- 장바구니 추가 method 구현 시 update_or_create 활용하여 해당 상품이 장바구니에 이미 존재하는지 확인해보려했으나
수량을 +1 update 혹은 1로 create 하는 로직을 구분하지 못했는데 가능한지 궁금합니다

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [ ] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

Commit:
ADD: 장바구니 구현
- 장바구니 추가, 삭제, 갯수 추가 view 1차 작성

Commit:
ADD: 장바구니 구현
- 장바구니 추가, 삭제, 갯수 추가 view 2차 작성

Commit:
ADD: 장바구니 구현
- 장바구니 추가, 삭제, 갯수 추가 view 3차 작성

Commit:
commit

Commit:
임시 commit

Commit:
장바구니 / 구매 전환 1차 구현

Commit:
ADD: 장바구니 구현
- 장바구니 추가, 삭제, 갯수, 조회 METHOD 구현
get_or_created 활용한 refactoring 진행
정상 구동 확인

Commit:
commit

Commit:
임시commit

Kang-kyunghun commented on CSV/category.csv:
```diff
@@ -0,0 +1,26 @@
+category,name,summary,thumbnail_url,description,content,content_image_url,count
```
csv 파일은 제거해 주세요.

Room9 replied:
네 알겠습니다!

Kang-kyunghun commented on db_update.py:
```diff
@@ -0,0 +1,86 @@
+import os
```
updata 파일도 제거하고 gitignor에 추가해 주세요.

Room9 replied:
수정완료했습니다

Kang-kyunghun commented on orders/views.py:
```diff
@@ -1,3 +1,114 @@
+from django.http            import JsonResponse
+from django.views           import View
+
+from orders.models   import WishList, OrderList, Order, PaymentMethod, OrderStatus
+from products.models import Product, ProductSelection
+from users.models    import User
+from my_settings     import SECRET
+
+from users.utils     import Authorization_decorator
```
import 라인 맞춰주세요.

Room9 replied:
수정 완료했습니다

Kang-kyunghun commented on orders/views.py:
```diff
@@ -1,3 +1,114 @@
+            quantity = 1
+        )        
+
+        # order_id = Order.objects.get(status_id=status_id).id
+
+        # orderlist, updated = OrderList.objects.update_or_create(product_selection_id=selection.id, default={
+        #             'order_id' : order_id,
+        #             'product_selection_id' : selection.id,
+        #             'quantity' : # 카트 내 상품이 기존재하는 경우 +1만 하려면 어떻게 표현할 수 있을까요? 
+        # })
+        
```
주석은 제거해 주세요.

Room9 replied:
수정 완료했습니다

Kang-kyunghun commented on orders/views.py:
```diff
@@ -1,3 +1,114 @@
+    @Authorization_decorator
+    def get(self, request):
+        try:
+            cartlists   = OrderList.objects.all() 
```
이렇게 되면 모든 orderlist가 다 담깁니다. 장바구니 목록을 보여주기 위한 view라고 하면 어떤 user의 것이고, 상태가 주문전인 orderlist을 가져와야 합니다.

Room9 replied:
user 구분 될 수 있도록 조건 추가했습니다

Kang-kyunghun commented on orders/views.py:
```diff
@@ -1,3 +1,114 @@
+            result = []
+            
+            for cartlist in cartlists:
+                selection_id = cartlist.product_selection_id
+                product_id   = ProductSelection.objects.get(id=selection_id).product_id
+                order_id     = cartlist.order_id
+                status_id    = Order.objects.get(id=order_id).status_id
+
+                cart_dict = {
+                    'name'    :Product.objects.get(id=product_id).name,
+                    'size'    :ProductSelection.objects.get(id=selection_id).size,
+                    'quantity':cartlist.quantity,
+                    'price'   :ProductSelection.objects.get(id=selection_id).price,
+                    'added_at':Order.objects.get(status_id=status_id).purchased_at
```
orderlist 데이터를 가져왔으면 그안에 product 정보들도 다 포함 되어 있을 것으로 예상됩니다. 가져온 orderlist로부터 데이터를 가져와 보세요.

Room9 replied:
orderlist 테이블이 order와 productselection을 참조하기만하여, 정참조 이용하여 해당 data 가져왔습니다 수정 방안 내일 재문의 드리겠습니다

Kang-kyunghun commented on orders/views.py:
```diff
@@ -1,3 +1,114 @@
+
+class CartView(View):
+    @Authorization_decorator
+    def post(self, request):
```
생선 부분은 구두로  이야기 하겠습니다.

Room9 replied:
넵 우선 user_id 구분 추가했습니다 내일 재 문의 드리겠습니다.

Commit:
refactor: 피드백 수정
get/filter에 user_id 조건 추가하여 user 별로 장바구니/order 구분 가능하도록 수정했습니다
csv 파일 삭제 및 update.py gitignore 추가하여 숨김처리했습니다

Commit:
refactor: 리뷰 사항 반영

Commit:
commit

Commit:
commit

Commit:
ADD: 주문내역 조회 구현 추가

Commit:
카트 구매

Commit:
conflict해결

Kang-kyunghun commented on orders/views.py:
```diff
@@ -0,0 +1,115 @@
+            cartlist.quantity += 1
+            cartlist.save()
+        else:
+            OrderList.objects.create(
+            order_id = Order.objects.get(status_id=status_id, user_id=user_id).id,
+            product_selection_id= selection.id,
+            quantity = 1
+        )        
```
```suggestion
            OrderList.objects.create(
            order_id = order.id,
            product_selection_id= selection.id,
            quantity = 1
        )        
```

Kang-kyunghun commented on orders/views.py:
```diff
@@ -0,0 +1,115 @@
+        data      = json.loads(request.body)
+
+        selection = ProductSelection.objects.get(product_id = data['product_id'], size = data['size'])
+        status_id = OrderStatus.objects.get(name='주문 전').id
```
```suggestion
        status = OrderStatus.objects.get(name='주문 전')
```

Kang-kyunghun commented on orders/views.py:
```diff
@@ -0,0 +1,115 @@
+        status_id = OrderStatus.objects.get(name='주문 전').id
+        user_id   = request.user.id
+        
+        order, created = Order.objects.get_or_create(status_id=status_id, user_id=user_id, defaults={
```
```suggestion
        order, created = Order.objects.get_or_create(status=status, user_id=user_id, defaults={
```

Commit:
conflict해결

Commit:
장바구니 refactoring

Kang-kyunghun commented:
LGTM
